File: hydra-scala/src/main/scala/hydra/lib/maybes.scala
package hydra.lib

object maybes:
  def apply[A, B](mf: Option[A => B])(ma: Option[A]): Option[B] = mf.flatMap(f => ma.map(f))
  def bind[A, B](ma: Option[A])(f: A => Option[B]): Option[B] = ma.flatMap(f)
  def cases[A, B](ma: Option[A])(ifNone: => B)(ifSome: A => B): B = ma match
    case None => ifNone
    case Some(a) => ifSome(a)
  def cat[A](xs: Seq[Option[A]]): Seq[A] = xs.flatten
  def compose[A, B, C](f: A => Option[B])(g: B => Option[C])(a: A): Option[C] = f(a).flatMap(g)
  def fromJust[A](ma: Option[A]): A = ma.get
  def fromMaybe[A](d: => A)(ma: Option[A]): A = ma.getOrElse(d)
  def isJust[A](ma: Option[A]): Boolean = ma.isDefined
  def isNothing[A](ma: Option[A]): Boolean = ma.isEmpty
  def map[A, B](f: A => B)(ma: Option[A]): Option[B] = ma.map(f)
  def mapMaybe[A, B](f: A => Option[B])(xs: Seq[A]): Seq[B] = xs.flatMap(f)
  def maybe[B, A](d: => B)(f: A => B)(ma: Option[A]): B = ma.map(f).getOrElse(d)
  def pure[A](a: A): Option[A] = Some(a)
  def toList[A](ma: Option[A]): Seq[A] = ma.toSeq

File: packages/hydra-rust/src/hydra/lib/maybes.rs
// hydra.lib.maybes primitives

pub fn apply<A, B>(mf: Option<std::rc::Rc<dyn Fn(A) -> B>>, ma: Option<A>) -> Option<B> {
    match (mf, ma) {
        (Some(f), Some(a)) => Some(f(a)),
        _ => None,
    }
}

pub fn bind<A, B>(m: Option<A>, f: impl Fn(A) -> Option<B>) -> Option<B> {
    m.and_then(|x| f(x))
}

pub fn cat<A>(l: Vec<Option<A>>) -> Vec<A> {
    l.into_iter().flatten().collect()
}

pub fn from_maybe<A>(default: A, m: Option<A>) -> A {
    m.unwrap_or(default)
}

pub fn is_just<A>(m: Option<A>) -> bool {
    m.is_some()
}

pub fn is_nothing<A>(m: Option<A>) -> bool {
    m.is_none()
}

pub fn map<A, B>(f: impl Fn(A) -> B, m: Option<A>) -> Option<B> {
    m.map(f)
}

pub fn maybe<A, B>(default: B, f: impl Fn(A) -> B, m: Option<A>) -> B {
    match m {
        Some(a) => f(a),
        None => default,
    }
}

pub fn pure_<A>(x: A) -> Option<A> {
    Some(x)
}
